File: se/se6029/mob-programming/self-practice/01-guinea-pig-car-car-cpp/src/guinea_pig_car_car.cpp
// guinea_pig_car_car.cpp

#include "guinea_pig_car_car.h"
#include <iostream>

GuineaPigCarCar::GuineaPigCarCar() : GuineaPig("GuineaPigCarCar"), car(), foods() {}

GuineaPigCarCar::GuineaPigCarCar(std::string name, Car car) : GuineaPig(name), car(car), foods() {}

void GuineaPigCarCar::eat(std::string food) {
    std::cout << name << " eat " << food << std::endl;
    eatRule(food);
}

void GuineaPigCarCar::pupu() {
    std::cout << name;
    for (const auto &pupu : foods) {
        std::cout << " " << pupu;
    }
    std::cout << std::endl;
    foods.clear();
    foodCount = 0;

    car.resetSpeed();
}

int GuineaPigCarCar::getSpeed() {
    return car.getSpeed();
}

void GuineaPigCarCar::eatRule(std::string food) {
    foods.push_back(food);

    if (foods.size() > 5) {
        car.resetSpeed();
    } else if (food == "carrot") {
        car.accelerate();
    } else {
        // do nothing
    }
}

File: se/se6029/mob-programming/01-guinea-pig-car-car/src/GuineaPig/GuineaPigCarCar.java
// GuineaPigCarCar.java

package GuineaPig;

import java.util.ArrayList;
import Car.Car;

public class GuineaPigCarCar extends GuineaPig {
    private Car car;
    private ArrayList<String> foods = new ArrayList<String>();

    public GuineaPigCarCar() {
        this("GuineaPigCarCar", new Car());
    }

    public GuineaPigCarCar(String name, Car car) {
        super(name);
        this.car = car;
    }

    public void eat(String food) {
        System.out.print(super.name + " eat " + food + ".");

        eatRule(food);
    }

    public void pupu() {
        // System.out.println(super.name + "pupu.");
        // 吃什麼棒什麼
        System.out.printf(super.name);
        for (String pupu : foods) {
            System.out.printf(" %s", pupu);
        }
   
        System.out.println();
        foods.clear();
        super.foodCount = 0;

        car.resetSpeed();
    }

    public int getSpeed() {
        return car.getSpeed();
    }

    // 這邊可以做到當規則變了修改這裡就好，或許可以寫個 Rule.java 來統一管理
    private void eatRule(String food) {
        foods.add(food);

        if (foods.size() > 5) {
            car.resetSpeed();
        } else if (food == "carrot") {
            car.accelerate();
        } else {
            // do nothing
        }
    }

}
